scrapy/core/spidermw.py

```diff
@@ -50,21 +55,36 @@ def process_spider_input(response):
 
         def process_spider_exception(_failure):
             exception = _failure.value
+            # ignore AssertionError from middleware's manager
+            if isinstance(exception, AssertionError) and len(exception.args) > 0 \
```

Comment:
This line checks for an AssertionError explicitly raised by any process_spider_input method, and prevents it from being passed to the exception handler (the only problem would be an external exception raised with the exact same message the spider middleware manager specifies, I believe that case could be ignored).